rmw_fastrtps_shared_cpp/src/custom_publisher_info.cpp

```diff
@@ -23,15 +24,74 @@
 EventListenerInterface *
 CustomPublisherInfo::get_listener() const
 {
-  return listener_;
+  return publisher_event_;
 }
 
-eprosima::fastdds::dds::StatusCondition & PubListener::get_statuscondition() const
+CustomDataWriterListener::CustomDataWriterListener(RMWPublisherEvent * pub_event)
+: publisher_event_(pub_event)
+{
+}
+
+void CustomDataWriterListener::on_publication_matched(
+  eprosima::fastdds::dds::DataWriter * writer,
+  const eprosima::fastdds::dds::PublicationMatchedStatus & status)
+{
+  (void)writer;
+
+  if (status.current_count_change == 1) {
+    publisher_event_->add_subscription(
+      eprosima::fastrtps::rtps::iHandle2GUID(status.last_subscription_handle));
+  } else if (status.current_count_change == -1) {
+    publisher_event_->remove_subscription(
+      eprosima::fastrtps::rtps::iHandle2GUID(status.last_subscription_handle));
+  }
+}
+
+
+void
+CustomDataWriterListener::on_offered_deadline_missed(
+  eprosima::fastdds::dds::DataWriter * writer,
+  const eprosima::fastdds::dds::OfferedDeadlineMissedStatus & status)
+{
+  (void)writer;
+
+  publisher_event_->update_deadline(status.total_count, status.total_count_change);
+}
+
+void CustomDataWriterListener::on_liveliness_lost(
+  eprosima::fastdds::dds::DataWriter * writer,
+  const eprosima::fastdds::dds::LivelinessLostStatus & status)
+{
+  (void)writer;
+
+  publisher_event_->update_liveliness_lost(status.total_count, status.total_count_change);
+}
+
+void CustomDataWriterListener::on_offered_incompatible_qos(
+  eprosima::fastdds::dds::DataWriter * writer,
+  const eprosima::fastdds::dds::OfferedIncompatibleQosStatus & status)
+{
+  (void)writer;
+
+  publisher_event_->update_offered_incompatible_qos(
+    status.last_policy_id, status.total_count,
+    status.total_count_change);
+}
+
+RMWPublisherEvent::RMWPublisherEvent(CustomPublisherInfo * info)
+: publisher_info_(info),
+  deadline_changed_(false),
+  liveliness_changed_(false),
+  incompatible_qos_changed_(false)
+{
+}
+
+eprosima::fastdds::dds::StatusCondition & RMWPublisherEvent::get_statuscondition() const
 {
   return publisher_info_->data_writer_->get_statuscondition();
 }
 
-bool PubListener::take_event(
+bool RMWPublisherEvent::take_event(
```

Comment:
It might just be me but I find it confusing that `RMWPublisherEvent` would be able to take events (I think of the event as a struct of information that can be manipulated, as opposed to having any listening capabilities, which is what seems to be happening in the body of this method.)

I'm wondering if there's a better name for the class. Or it might just be a bad name for the method in the current version of the library, or I might just be confusing myself.

Perhaps `update_event`?

If changes are made addressing this comment, they should be rolled out to the subscriber and event info too.

Reply:
So what the `RMWPublisherEvent` class does is to collect data about events that happened to the publisher, and make them available to the RMW layer.  In this new refactor, this class doesn't directly register for any event callbacks; that is done via the `CustomDataWriterListener` and `CustomTopicListener`, and when the callbacks fire they call their saved handle to this class to update the data.  Then the layers about the RMW get notified of the event, and can come fetch the event via `take_event`.

So to me, the `take_event` makes sense; the upper layers are taking the data for this event.  The name of the class is more up-in-the-air; it really is a collection of publisher events collected for the RMW layer, hence the current name.  But I'm open to naming it something different, if you have a better idea.  Let me know what you think.